File: remote/src/main/java/com/kelsos/mbrc/ui/fragments/MiniControlFragment.java
package com.kelsos.mbrc.ui.fragments;

import android.graphics.Typeface;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;
import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;
import com.google.inject.Inject;
import com.kelsos.mbrc.R;
import com.kelsos.mbrc.constants.Protocol;
import com.kelsos.mbrc.constants.ProtocolEventType;
import com.kelsos.mbrc.data.UserAction;
import com.kelsos.mbrc.enums.DisplayFragment;
import com.kelsos.mbrc.events.MessageEvent;
import com.kelsos.mbrc.events.ui.CoverAvailable;
import com.kelsos.mbrc.events.ui.DrawerEvent;
import com.kelsos.mbrc.events.ui.PlayStateChange;
import com.kelsos.mbrc.events.ui.TrackInfoChange;
import com.squareup.otto.Bus;
import com.squareup.otto.Subscribe;
import roboguice.fragment.RoboFragment;

public class MiniControlFragment extends RoboFragment {

  @Inject Bus bus;
  @Bind(R.id.mc_track_cover) ImageView trackCover;
  @Bind(R.id.mc_track_artist) TextView trackArtist;
  @Bind(R.id.mc_track_title) TextView trackTitle;
  @Bind(R.id.mc_play_pause) ImageButton playPause;

  @OnClick(R.id.mini_control) public void onControlClick() {
    bus.post(new DrawerEvent(DisplayFragment.HOME, true));
  }

  @OnClick(R.id.mc_next_track) public void onNextClick(View view) {
    bus.post(
        new MessageEvent(ProtocolEventType.UserAction, new UserAction(Protocol.PlayerNext, true)));
  }

  @OnClick(R.id.mc_play_pause) public void onPlayClick(View view) {
    bus.post(new MessageEvent(ProtocolEventType.UserAction,
        new UserAction(Protocol.PlayerPlayPause, true)));
  }

  @OnClick(R.id.mc_prev_track) public void onPreviousClick(View view) {
    bus.post(new MessageEvent(ProtocolEventType.UserAction,
        new UserAction(Protocol.PlayerPrevious, true)));
  }

  @Override public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
  }

  @Override
  public View onCreateView(LayoutInflater inflater, ViewGroup container,
      Bundle savedInstanceState) {
    final View view = inflater.inflate(R.layout.ui_fragment_mini_control, container, false);
    ButterKnife.bind(this, view);
    return view;
  }

  @Override public void onStart() {
    super.onStart();
    bus.register(this);
    Typeface robotoRegular =
        Typeface.createFromAsset(getActivity().getAssets(), "fonts/roboto_regular.ttf");
    Typeface robotoMedium =
        Typeface.createFromAsset(getActivity().getAssets(), "fonts/roboto_medium.ttf");
    trackTitle.setTypeface(robotoMedium);
    trackArtist.setTypeface(robotoRegular);
  }

  @Override public void onStop() {
    super.onStop();
    bus.unregister(this);
  }

  @Subscribe public void handleCoverChange(CoverAvailable event) {
    if (trackCover == null) {
      return;
    }
    if (event.isAvailable()) {
      trackCover.setImageBitmap(event.getCover());
    } else {
      trackCover.setImageResource(R.drawable.ic_image_no_cover);
    }
  }

  @Subscribe public void handleTrackInfoChange(TrackInfoChange event) {
    trackArtist.setText(event.getArtist());
    trackTitle.setText(event.getTitle());
  }

  @Subscribe public void handlePlayStateChange(PlayStateChange event) {
    switch (event.getState()) {
      case Playing:
        playPause.setImageResource(R.drawable.ic_action_pause);
        break;
      default:
        playPause.setImageResource(R.drawable.ic_action_play);
        break;
    }
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/ui/fragments/MiniControlFragment.kt
package com.kelsos.mbrc.ui.fragments

import android.graphics.Bitmap
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import butterknife.BindView
import butterknife.ButterKnife
import butterknife.OnClick
import javax.inject.Inject
import com.kelsos.mbrc.R
import com.kelsos.mbrc.annotations.PlayerState
import com.kelsos.mbrc.presenters.MiniControlPresenter
import com.kelsos.mbrc.ui.views.MiniControlView
import com.kelsos.mbrc.utilities.FontUtils
import roboguice.RoboGuice

class MiniControlFragment : Fragment(), MiniControlView {

  @BindView(R.id.mc_track_cover) internal lateinit var trackCover: ImageView
  @BindView(R.id.mc_track_artist) internal lateinit var trackArtist: TextView
  @BindView(R.id.mc_track_title) internal lateinit var trackTitle: TextView
  @BindView(R.id.mc_play_pause) internal lateinit var playPause: ImageButton

  @Inject private lateinit var presenter: MiniControlPresenter

  @OnClick(R.id.mc_next_track) fun onNextClicked() {
    presenter.onNextPressed()
  }

  @OnClick(R.id.mc_play_pause) fun onPlayPauseClicked() {
    presenter.onPlayPause()
  }

  @OnClick(R.id.mc_prev_track) fun onPreviousClicked() {
    presenter.onPreviousPressed()
  }

  override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
    val view = inflater!!.inflate(R.layout.ui_fragment_mini_control, container, false)
    ButterKnife.bind(this, view)
    RoboGuice.getInjector(context).injectMembers(this)
    presenter.bind(this)
    trackTitle.typeface = FontUtils.getRobotoMedium(activity)
    trackArtist.typeface = FontUtils.getRobotoRegular(activity)
    presenter.load()
    return view
  }

  override fun onPause() {
    super.onPause()
    presenter.onPause()
  }

  override fun onResume() {
    super.onResume()
    presenter.onResume()
  }

  override fun updateCover(cover: Bitmap?) {
    if (cover != null) {
      trackCover.setImageBitmap(cover)
    } else {
      trackCover.setImageResource(R.drawable.ic_image_no_cover)
    }
  }

  override fun updateTrack(artist: String, title: String) {
    trackArtist.text = artist
    trackTitle.text = title
  }

  override fun updatePlayerState(@PlayerState.State state: String) {
    if (state == PlayerState.PLAYING) {
      playPause.setImageResource(R.drawable.ic_pause_black_36dp)
    }
    else {
      playPause.setImageResource(R.drawable.ic_play_arrow_black_36dp)
    }
  }

  companion object {

    fun newInstance(): Fragment {
      return MiniControlFragment()
    }
  }
}
